stl/inc/algorithm

```diff
@@ -2797,6 +2797,220 @@ namespace ranges {
+
+    class _Find_last_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <forward_iterator _It, sentinel_for<_It> _Se, class _Ty, class _Pj = identity>
+            requires indirect_binary_predicate<ranges::equal_to, projected<_It, _Pj>, const _Ty*>
+        _NODISCARD constexpr subrange<_It> operator()(_It _First, _Se _Sent, const _Ty& _Value, _Pj _Proj = {}) const {
+            _Adl_verify_range(_First, _Sent);
+
+            auto [_UFirst, _ULast] = _Find_last_unchecked(
+                _Unwrap_iter<_Se>(_STD move(_First)), _Unwrap_sent<_It>(_STD move(_Sent)), _Value, _Pass_fn(_Proj));
+            _It _Last{};
+            _Seek_wrapped(_First, _STD move(_UFirst));
+            _Seek_wrapped(_Last, _STD move(_ULast));
+            return {_STD move(_First), _STD move(_Last)};
+        }
+
+        template <forward_range _Rng, class _Ty, class _Pj = identity>
+            requires indirect_binary_predicate<ranges::equal_to, projected<iterator_t<_Rng>, _Pj>, const _Ty*>
+        _NODISCARD constexpr borrowed_subrange_t<_Rng> operator()(
+            _Rng&& _Range, const _Ty& _Value, _Pj _Proj = {}) const {
+            auto _First            = _RANGES begin(_Range);
+            auto [_UFirst, _ULast] = _Find_last_unchecked(
+                _Unwrap_range_iter<_Rng>(_STD move(_First)), _Uend(_Range), _Value, _Pass_fn(_Proj));
+            iterator_t<_Rng> _Last{};
+            _Seek_wrapped(_First, _STD move(_UFirst));
+            _Seek_wrapped(_Last, _STD move(_ULast));
+            return {_STD move(_First), _STD move(_Last)};
+        }
+
+    private:
+        template <class _It, class _Se, class _Ty, class _Pj>
+        _NODISCARD static constexpr subrange<_It> _Find_last_unchecked(
+            _It _First, _Se _Last, const _Ty& _Value, _Pj _Proj) {
+            _STL_INTERNAL_STATIC_ASSERT(forward_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+            _STL_INTERNAL_STATIC_ASSERT(indirect_binary_predicate<ranges::equal_to, projected<_It, _Pj>, const _Ty*>);
+
+            if constexpr (_Bidi_common<_It, _Se>) {
+                for (auto _Result = _Last; _Result != _First;) {
+                    if (_STD invoke(_Proj, *--_Result) == _Value) {
+                        return {_STD move(_Result), _STD move(_Last)};
+                    }
+                }
+                return {_Last, _Last};
+            } else if constexpr (same_as<_It, _Se>) {
+                auto _Result = _Last;
+                for (; _First != _Last; ++_First) {
+                    if (_STD invoke(_Proj, *_First) == _Value) {
+                        _Result = _First;
+                    }
+                }
+                return {_STD move(_Result), _STD move(_Last)};
+            } else {
+                _It _Result{};
+                for (;;) {
+                    if (_STD invoke(_Proj, *_First) == _Value) {

```

Comment:
Several issues here, all related:

* This is dereferencing `_First`, but the range could be empty.
* When we don't find a result (e.g. because the range is empty, or it's non-empty but nothing matched), we need to return `{last, last}`. There's code below that attempts to achieve this, with `if (_Result == _It{})`, but that is not allowed by the iterator requirements. Value-initialized iterators are comparable (and guaranteed equal) to other value-initialized iterators, but they *cannot* be compared to other iterators. (In practice, STL iterators won't detect this issue, because they participate in unwrapping, so by the time we reach this code, the debug checks that would detect it have been unwrapped away. However, the Standard's requirements still apply, and it is totally possible for user-defined iterators to lack any unwrapping machinery and simultaneously enforce the rule against comparing value-initialized with parented iterators.)
* This indicates a lack of test coverage for empty ranges (forward non-common in this case).